src/v/datalake/partition_spec_parser.cc

```diff
@@ -10,260 +10,516 @@
+    parse_result<std::monostate> expect_with_ws(position pos) {
+        skip_ws(pos);
+
+        source unparsed{pos, end()};
+        constexpr std::conditional_t<case_sensitive, std::identity, upper> proj;
+        if (!std::ranges::starts_with(
+              unparsed, std::string_view{expected}, {}, proj, proj)) {
+            static constexpr auto err = fixed_string{"\""} + expected
+                                        + fixed_string{"\""};
+            return {*this, err, pos};
+        }
+        pos += std::string_view{expected}.size();
 
-    auto transform_id = parse_identifier(unparsed);
-    if (!transform_id) {
-        return std::nullopt;
+        skip_ws(pos);
+
+        return {std::monostate{}, pos};
     }
-    iceberg::transform transform;
-    if (transform_id->val == "hour") {
-        transform = iceberg::hour_transform{};
-    } else if (transform_id->val == "day") {
-        transform = iceberg::day_transform{};
-    } else if (transform_id->val == "identity") {
-        transform = iceberg::identity_transform{};
-    } else {
-        // TODO: parse all transforms
-        ctx.report_expected(unparsed, "known transform name");
-        return std::nullopt;
+
+    parse_result<ss::sstring> parse_simple_identifier(position pos) {
+        // first symbol cannot be a digit
+        constexpr static auto expected1 = "letter or \"_\"";
+        if (pos == end() || (*pos != '_' && !std::isalpha(*pos))) {
+            return {*this, expected1, pos};
+        }
+        // remaining symbols
+        position cur = pos + 1;
+        constexpr static auto expected = "alphanumeric or \"_\"";
+        while (cur != end() && (*cur == '_' || std::isalnum(*cur))) {
+            ++cur;
+        }
+        ss::sstring id{pos, cur};
+        for (char& c : id) {
+            c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
+        }
+        return {std::move(id), cur, *this, expected, cur};
     }
-    unparsed = transform_id->unparsed;
 
-    skip_space(unparsed);
-    if (!skip_expected(unparsed, "(")) {
-        ctx.report_expected(unparsed, "'('");
-        return std::nullopt;
+    template<char quotation_mark>
+    parse_result<ss::sstring> parse_quoted_w_doubling(position pos) {
+        constexpr char quotation_mark_as_array[2]{quotation_mark, '\0'};
+        auto o_q_res = expect_with_ws<quotation_mark_as_array>(pos);
+        if (!o_q_res) {
+            return o_q_res;
+        }
+
+        pos = o_q_res.after();
+        position cur = pos;
+        while (cur != end()) {
+            if (*cur == quotation_mark) {
+                ++cur;
+                auto dbl_q_res = expect_with_ws<quotation_mark_as_array>(cur);
+                if (!dbl_q_res) {
+                    // it was a single quote: finalize!
+                    ss::sstring res{};
+                    position cur2 = pos;
+                    while (cur2 != cur) {
+                        res.append(&*cur2, 1);
```

Comment:
can it be a simple push_back?

Reply:
is there such a method in sstring?

Reply:
looks like append() copies the whole string...

Reply:
changed to pre-allocate required length